### ISSUE TITLE
[css-overflow-3] What clips scrollable overflow?

### ISSUE THREAD

**Issue by fantasai** [w3c/csswg-drafts#8607]
See [testcase](https://www.software.hixie.ch/utilities/js/live-dom-viewer/?%3C!DOCTYPE%20html%3E%0A%3Cstyle%3E%0A%20%20.scroller%20%7B%20width%3A%2060px%3Bheight%3A%2060px%3B%20overflow%3A%20scroll%3B%20border%3A%20solid%3B%20margin%3A%201em%3B%20float%3A%20left%3B%20%7D%0A%20%20.clipper%20%7B%20width%3A%2010px%3B%20height%3A%2010px%20%7D%0A%20%20.clippee%20%7B%20border%3A%20100px%20orange%20solid%20%7D%0A%3C%2Fstyle%3E%0A%3Cdiv%20class%3D%22scroller%22%3E%0A%20%20%3Cdiv%20class%3D%22clipper%22%20style%3D%22overflow%3A%20clip%3B%22%3E%0A%20%20%20%20%3Cdiv%20class%3D%22clippee%22%3E%3C%2Fdiv%3E%0A%20%20%3C%2Fdiv%3E%0A%3C%2Fdiv%3E%0A%3Cdiv%20class%3D%22scroller%22%3E%0A%20%20%3Cdiv%20class%3D%22clipper%22%20style%3D%22contain%3A%20paint%22%3E%0A%20%20%20%20%3Cdiv%20class%3D%22clippee%22%3E%3C%2Fdiv%3E%0A%20%20%3C%2Fdiv%3E%0A%3C%2Fdiv%3E%0A%3Cdiv%20class%3D%22scroller%22%3E%0A%20%20%3Cdiv%20class%3D%22clipper%22%20style%3D%22clip-path%3A%20inset(0%200%200%200)%22%3E%0A%20%20%20%20%3Cdiv%20class%3D%22clippee%22%3E%3C%2Fdiv%3E%0A%20%20%3C%2Fdiv%3E%0A%3C%2Fdiv%3E%0A%3Cdiv%20class%3D%22scroller%22%20style%3D%22position%3A%20relative%22%3E%0A%20%20%3Cdiv%20class%3D%22clipper%22%20style%3D%22position%3A%20absolute%3B%20width%3A%20auto%3B%20clip%3A%20rect(0%2C10px%2C10px%2C0)%3B%22%3E%0A%20%20%20%20%3Cdiv%20class%3D%22clippee%22%3E%3C%2Fdiv%3E%0A%20%20%3C%2Fdiv%3E%0A%3C%2Fdiv%3E%0A):

So far the following results:
* `overflow: clip` and `contain: paint` clip scrollable overflow
* `clip-path` does not clip scrollable overflow in FF, WK, or Blink
    * Side note: Blink doesn't seem to support the full syntax of `clip-path`, e.g. `border-box` doesn't work >:/
* `clip` does not have interop on this point

Is this the behavior we want?

**Comment by smfr** [comment]
`clip-path` is like a mask or filter, which does not affect layout overflow. `clip` is just weird and I don't really care :)

**Comment by bfgeek** [comment]
I believe we represent `clip` and `clip-path` as the same concept internally - so I'd prefer if they were kept consistent.

**Comment by frivoal** [comment]
I think we have a little less interop than we thought with `overflow: clip`.

TL;DR: The problematic case is when a float overflows in the block axis.

Here's a [test case](http://software.hixie.ch/utilities/js/live-dom-viewer/?saved=11610) It has a float, inside of a box with `overflow: clip`, itself followed by a box with `clear: both`, all of that inside of a box with `overflow: auto`. Then that's repeated, but without the box with clearance.

First, in Safari, the test doesn't work, because it seems that `overflow: clip` stops having any effect at all (!) if a float overflows in the block axis. It would take effect in the inline axis if there was no block overflow, or if any block axis overflow was caused by something other than a float, like an abspos. Hard to think that's anything but a bug, but that bug could be a sign of "it's harder than we thought". [testcase for that](http://software.hixie.ch/utilities/js/live-dom-viewer/?saved=11611)

Back to the main test case.

In Chrome and Firefox (and Safari if you modify the test case to remove block overflow), `overflow: clip` does suppress the inline axis scrolling potentially caused by a protruding float. This is what the opening comment tested for and claimed interop about.

However, things get more "interesting" in the block axis:
* As stated above, `overflow: clip` stops working altogether in Safari when a float overflows in the block axis
* In both Chrome and Firefox, there is interop about the spec requirement that an `overflow: clip` element does not establish a BFC, and therefore the box with clearance is pushed down by the invisible geometry of the clipped float. Therefore we do get vertical scrolling (triggered by the box with clearance).
* However, if we remove the box with clearance, we get a difference:
    *  In Chrome, just like in the inline axis, the invisible geometry of the clipped float **does not** trigger scroll overflow in the block axis, so we **do not** get a vertical scrollbar.
    *  In Firefox, the invisible geometry of the clipped float **does** trigger scroll overflow in the block axis, so we **do** get a vertical scrollbar.

If we try to have something other than a float cause the overflow, [for instance with an abspos](http://software.hixie.ch/utilities/js/live-dom-viewer/?saved=11613), then we do have interop in all 3 browsers that the clipped content **does not** trigger scroll overflow in either axis.

----

In conclusion, what the spec currently says is that `overflow:clip`  visually clips the content without providing scrollbars (nor programatic scrolling) on the clipping element itself. I believe we do have interop on that.

But whether the protruding geometry can cause scrollbars on ancestor elements does not have full interop. We seem closer to agreement on not scrolling than the other way around, but we're not fully there: when a float overflows in the block axis, we get 3 possible behaviors: clear and scroll, clear but don't scroll, or give up.

Question to the group (and particularly to Safari and Firefox): do you consider that a bug that you intend to fix, and we can therefore stick with the resolution, or does that cause a need to rethink?

**Comment by dholbert** [comment]
> I think we have a little less interop than we thought with `overflow: clip`.
> [...]
> Question to the group (and particularly to Safari and Firefox): do you consider that a bug that you intend to fix, and we can therefore stick with the resolution, or does that cause a need to rethink?

I think the Firefox behavior there is probably a bug -- at least, we change to match Chromium if I apply @dshin-moz's patch for [bug 1768921](https://bugzilla.mozilla.org/show_bug.cgi?id=1768921) (which is aiming to implement the spec update from #129 and align with Chromium's behavior on that).

(That patch has been stuck in my review queue, but I anticipate finishing the review soon.  When it lands, Firefox & Chromium will be consistent on this part at least.)

So: no concerns from me, on sticking with the resolution at this point.

**Comment by smfr** [comment]
I filed https://bugs.webkit.org/show_bug.cgi?id=256410 on the Safari issue with floats.
